src/core/ee/emotion_breakpoint.cpp

```diff
@@ -0,0 +1,745 @@
+
+    for(;i1 != breakpoints.end();)
+    {
+        // get rid of breakpoint and update iterators
+        if(*i2 && (*i1)->clear_after_hit())
+        {
+            delete *i1;
+            i1 = breakpoints.erase(i1);
+            i2 = breakpoint_triggered.erase(i2);
+        }
+        else
+        {
+            i1++;
+            i2++;
+        }
+    }
+
+    if(breakpoints.size() != breakpoint_triggered.size())
+        Errors::die("clear_old_breakpoints size error after clearing");
+
+    if(breakpoints.empty())
+        debug_enable = false;
+
+    unlock_list_mutex();
+}
+
+void BreakpointList::send_continue()
+{
+    for (auto &&i : breakpoint_triggered)
+        i = false;
+    break_cv.notify_all();
+}
+
+// return a list of all memory addresses which are breakpoint related
+// will recursively grab breakpoints from and
+std::vector<uint32_t> BreakpointList::get_memory_address_list()
+{
+    std::vector<uint32_t> v;
+    for(auto* bp : breakpoints)
+        append_memory_address_list(bp, v);
+
+    return v;
+}
+
+void BreakpointList::append_memory_address_list(DebuggerBreakpoint *bp, std::vector<uint32_t> &v)
+{
+    auto* and_breakpoint = dynamic_cast<EmotionBreakpoints::And*>(bp);
+    if(and_breakpoint)
+    {
+        append_memory_address_list(and_breakpoint->_a, v);
+        append_memory_address_list(and_breakpoint->_b, v);
+    }
+    else
+    {
+        uint32_t addr = bp->get_memory_address();
+        if(addr != 0xffffffff) v.push_back(addr);
+    }
+}
+
+// return a list of all memory addresses which are breakpoint related
+// will recursively grab breakpoints from and
+std::vector<uint32_t> BreakpointList::get_instruction_address_list()
+{
+    std::vector<uint32_t> v;
+    for(auto* bp : breakpoints)
+        append_instruction_address_list(bp, v);
+
+    return v;
+}
+
+void BreakpointList::append_instruction_address_list(DebuggerBreakpoint *bp, std::vector<uint32_t> &v)
+{
+    auto* and_breakpoint = dynamic_cast<EmotionBreakpoints::And*>(bp);
+    if(and_breakpoint)
+    {
+        append_instruction_address_list(and_breakpoint->_a, v);
+        append_instruction_address_list(and_breakpoint->_b, v);
+    }
+    else
+    {
+        uint32_t addr = bp->get_instruction_address();
+        if(addr != 0xffffffff) v.push_back(addr);
+    }
+}
+
+std::vector<uint64_t> BreakpointList::get_breakpoints_at_instruction_addr(uint32_t addr)
+{
+    std::vector<uint64_t> v;
+    for(uint64_t i = 0; i < breakpoints.size(); i++)
+    {
+        if(breakpoint_uses_instruction_addr(breakpoints[i], addr))
+            v.push_back(i);
+    }
+    return v;
+}
+
+bool BreakpointList::breakpoint_uses_instruction_addr(DebuggerBreakpoint *bp, uint32_t addr)
+{
+    auto* and_breakpoint = dynamic_cast<EmotionBreakpoints::And*>(bp);
+    if(and_breakpoint)
+    {
+        return breakpoint_uses_instruction_addr(and_breakpoint->_a, addr) ||
+        breakpoint_uses_instruction_addr(and_breakpoint->_b, addr);
```

Comment:
Might want to indent this line just so it is obvious that it is part of the previous instruction.